app/controllers/subscriptions_controller.rb

```diff
@@ -28,6 +28,33 @@ def create
     render json: { id: subscription.id }, status: status
   end
 
+  def change_frequency
+    existing_subscription = nil
+    subscription = nil
+
+    Subscription.transaction do
+      existing_subscription = Subscription.active.lock.find(
+        subscription_params.require(:subscription_id)
+      )
+
+      existing_subscription.end(reason: :frequency_changed)
+
+      subscription = Subscription.create!(
+        subscriber: existing_subscription.subscriber,
+        subscriber_list: existing_subscription.subscriber_list,
+        frequency: frequency,
+        signon_user_uid: current_user.uid,
+        source: :frequency_changed
+      )
+    end
+
+    render json: { subscription: subscription }, status: :ok
+  rescue ArgumentError
```

Comment:
I think we should wrap this around the `create!` in case we get an `ArgumentError` on a different line. Also, I wonder if a lot of this code can be extracted out into separate methods since it's so similar to the `create` method.

Reply:
I've changed the `ArgumentError` rescue block to be more specific. I'm struggling a little to identify things worth extracting into their own method since things are just different enough to make the methods nearly identical to what they're replacing IMO. Do you have any specific things you think could be extracted?

Reply:
Hmm yeah, I had a quick go myself and there wasn't much benefit in it.